// bamboolib/src/code_export.js

/*
This modules contains the CodeExporter classes both for Jupyter Notebook and Lab.
Those classes are responsible for setting the text of a code cell as part of the bamboolib live code export.
*/

class CodeExporter {
    maybeSetupForNotebook(){
        if(typeof this.code_exporter === 'undefined'){
            this.setupJupyterNotebook()
        }
        // else JupyterLab has already been setup
    }

    setupJupyterLab(notebookTracker){
        this.code_exporter = new LabCodeExporter(notebookTracker)
    }

    setupJupyterNotebook(){
        this.code_exporter = new NotebookCodeExporter()
    }

    setValue($el, new_code){
        try {
            this.code_exporter.setValue($el, new_code)
        } catch (e) {
            // e.g. maybe within databricks where the CodeExporters dont work yet
        }
    }

    getValue($el){
        try {
            return this.code_exporter.getValue($el)
        } catch (e) {
            return ""
            // e.g. maybe within databricks where the CodeExporters dont work yet
        }
    }
}

class NotebookCodeExporter {
    _getCodeCell($el){
        var parent_cell_element = $el.parents(".cell")[0]
        var cell_idx = window.Jupyter.notebook.get_cell_elements().index(parent_cell_element);
        var cell = window.Jupyter.notebook.get_cell(cell_idx);
        return cell.code_mirror.getDoc()
    }

    setValue($el, new_code){
        this._getCodeCell($el).setValue(new_code)
    }

    getValue($el){
        return this._getCodeCell($el).getValue()
    }
}

class LabCodeExporter {
    constructor(notebookTracker){
        this.notebookTracker = notebookTracker
    }

    _getCodeCell($el){
        var notebook = $el.parents(".jp-Notebook")[0]
        var cell = $el.parents(".jp-Cell")[0]
        var index = $(notebook).children().index(cell)
        return this.notebookTracker.currentWidget.model.cells.get(index).value
    }

    setValue($el, new_code){
        this._getCodeCell($el).text = new_code
    }

    getValue($el){
        return this._getCodeCell($el).text
    }
}

module.exports = {
    CodeExporter: new CodeExporter(),
};
